02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/06 PyTorch/05 Train Models.html

```diff
@@ -0,0 +1,72 @@
+<p>You can train the model at the beginning of your algorithm and you can periodically re-train it as the algorithm executes.</p>
+
+<h4>Warm Up Training Data</h4>
+<p>You need historical data to initially train the model at the start of your algorithm. To get the initial training data, in the <code>Initialize</code> method, make a <a href="https://www.quantconnect.com/docs/v2/writing-algorithms/historical-data/history-requests">history request</a>.</p>
+<div class="section-example-container">
+    <pre class="python">training_length = 252*2
+self.training_data = RollingWindow[float](training_length)
+history = self.History[TradeBar](self.symbol, training_length, Resolution.Daily)
+for trade_bar in history:
+    self.training_data.Add(trade_bar.Close)</pre>
+</div>
+
+<h4>Define a Training Method</h4>
+<p>To train the model, define a method that fits the model with the training data.</p>
+<div class="section-example-container">
+    <pre class="python">def get_features_and_labels(self, n_steps=5):
+    training_df = list(self.training_data)[::-1]
```

Comment:
`training_df` <- I don't think this is a DataFrame, let's change the variable name